### ISSUE TITLE
Digital Touch Support

### ISSUE THREAD

**Issue by trymoose** [ReagentX/imessage-exporter#344]
When researching handwritten messages I found digital touch messages were handled in bluebubbles the same way [here](https://github.com/BlueBubblesApp/bluebubbles-server/blob/87679b70588719a7032d457e1b9bce672be7c04a/packages/server/src/server/databases/imessage/entity/Message.ts#L626) and [here](https://github.com/BlueBubblesApp/bluebubbles-server/blob/87679b70588719a7032d457e1b9bce672be7c04a/packages/server/src/server/api/interfaces/messageInterface.ts#L517). Piping them through `protoscope` gives a valid message but it is different and more complex than the handwritten ones.

**Comment by ReagentX** [comment]
~~Digital Touch is already supported—they are sent as gifs.~~

**Comment by ReagentX** [comment]
Actually this appears to have changed–they no longer get converted by the system. For what it's worth, they aren't supported by macOS either:

![image](https://github.com/user-attachments/assets/615cf0c9-4ff2-462f-af89-c9e4dcd4cb2f)

**Comment by ReagentX** [comment]
They also sometimes seem to not even have the requisite payload:

![image](https://github.com/user-attachments/assets/b5ae5705-c971-4895-874a-72141776b4be)

**Comment by trymoose** [comment]
I've been generating and viewing them on iOS. I don't think I've seen one with a `NULL` payload before.

**Comment by ReagentX** [comment]
Ok, so the one that had a `NULL` payload used the `subject` column (the one used for message subject lines) contained a GUID that mapped to an existing Digital Touch message in the conversation: 

![image](https://github.com/user-attachments/assets/8fd397df-1791-4f33-bb89-df15437cdf11)

**Comment by trymoose** [comment]
[I've made some progress implementing support for these messages.](https://github.com/trymoose/imessage-exporter/tree/feature/digital-touch)

- [x] Tap
- [x] Heartbeat
  - [x] Heartbreak
- [x] Kiss
- [x] Sketch
- [x] Fireball
- [ ] Photo
- [ ] Video

**Comment by trymoose** [comment]
> Ok, so the one that had a `NULL` payload used the `subject` column (the one used for message subject lines) contained a GUID that mapped to an existing Digital Touch message in the conversation:
> 
> ![image](https://private-user-images.githubusercontent.com/1920666/368822893-8fd397df-1791-4f33-bb89-df15437cdf11.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3MTQ4ODQsIm5iZiI6MTcyNjcxNDU4NCwicGF0aCI6Ii8xOTIwNjY2LzM2ODgyMjg5My04ZmQzOTdkZi0xNzkxLTRmMzMtYmI4OS1kZjE1NDM3Y2RmMTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTlUMDI1NjI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWY3MGQ5NGM3ZTc5YjNmMzRjNTIzYWUwOTBhODk1Y2M0NzZmMmVlOTQxZDM0N2MwNzgwMWU4MDBiZTMyM2I5NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.B4XNVMFgZxVpRnuwOYanXx6EV7KOKQgo_TlHr2S7qEs)

Was the `NULL` message copy and pasted in iMessage?

**Comment by ReagentX** [comment]
It was not pasted and also does not show up in the thread.

**Comment by trymoose** [comment]
The only other thought I have right now is that it might be related to the announcements that show up saying "person kept a Digital Touch Message from you."

**Comment by ReagentX** [comment]
~~Those seem to come from one of the date columns.~~

**Comment by ReagentX** [comment]
The `NULL` Digital Touch messages do not have `date_played` and also have a custom type of `5`, which may indicate the "Kept" status as you mentioned:

<img width="252" alt="image" src="https://github.com/user-attachments/assets/51788d9a-2e14-4c52-8339-0131f06c5284">

Also, they do not always have the subject field populated and contain no `typedstream`-encoded body data:

<img width="491" alt="image" src="https://github.com/user-attachments/assets/cdf26e77-bbcb-4194-9236-cd31008a461c">

They also set this `message_action_type` field and do not have expiry data:

<img width="364" alt="image" src="https://github.com/user-attachments/assets/e346fc25-0f5a-4f6d-a15f-d5b0a7e18bd5">

In my database, an `item_type` of `5` always correlates with a `message_action_type` of `2`.

**Comment by trymoose** [comment]
I found some of the `NULL` payloads in my databases. I saw the database shows they are not delivered but `finished`:

![Screenshot from 2024-09-19 20-23-50](https://github.com/user-attachments/assets/142826cc-1604-426c-ac78-e480bda4feb0)

I had the same `item_type` of 5 and `message_action_type` of 2.

**Comment by ReagentX** [comment]
Yes, I see the `finished` status too—it also appears for messages sent with full-screen expressives.

**Comment by ReagentX** [comment]
<img width="292" alt="image" src="https://github.com/user-attachments/assets/8035f8fb-711b-4ff0-805f-1deff1df3404">

Adding rudimentary support for displaying the type of message for now.

**Comment by ReagentX** [comment]
Media attachments are joined to Digital Touch messages by the `message_attachment_join` table, so at construction time, we can apply the background by querying for the media with `Attachment::from_message()`.

**Comment by ReagentX** [comment]
<img width="1440" height="1440" alt="Image" src="https://github.com/user-attachments/assets/1e916e47-e765-4698-980e-35655200efe4" />

Note: this is the raw SVG output (static, not animated, though frames are preserved) prior to any media query, which would happen at render time, depending on the export type.

**Comment by ReagentX** [comment]
<img width="1440" height="1440" alt="Image" src="https://github.com/user-attachments/assets/3362846b-e291-44b9-955c-7bfed173a692" />
